### PR TITLE
Create deleteCoreV1CollectionNamespacedPodTemplate test+promote - +1 endpoint coverage

### LINKED ISSUE #90912: Write deleteCoreV1CollectionNamespacedPodTemplate test+promote - +1 endpoint coverage
# Progress `[7/7]`
* [x] Original approval issue: #90746
* [x]  APISnoop org-flow: [deleteCoreV1CollectionNamespacedPodTemplate.org](https://github.com/cncf/apisnoop/blob/master/tickets/k8s/deleteCoreV1CollectionNamespacedPodTemplateTest.org)
* [x]  test approval issue: [kubernetes/kubernetes#90912](https://github.com/kubernetes/kubernetes/issues/90912)
* [x]  test pr : [#90912](https://github.com/kubernetes/kubernetes/pull/91259)
* [x]  two weeks soak start date: 17 Jun 2020 [testgrid-link](https://testgrid.k8s.io/sig-release-master-blocking#gce-cos-master-default&width=5&include-filter-by-regex=PodTemplates.should.delete.a.collection.of.pod.templates) 
* [x]  two weeks soak end date: 2 July 2020
* [x]  test promotion pr: [#92812](https://github.com/kubernetes/kubernetes/pull/92812)

This issue is created to allow edit by @Riaankl

# Identifying an untested feature Using APISnoop
According to this APIsnoop query, there are still an endpoint which is untested.

/wip
/hold

```
SELECT
  operation_id,
  -- k8s_action,
  -- path,
  description,
  kind
  FROM untested_stable_core_endpoints
  -- FROM untested_stable_endpoints
  where path not like '%volume%'
  -- and kind like ''
  and operation_id ilike 'deleteCoreV1CollectionNamespacedPodTemplate'
 ORDER BY kind,operation_id desc
 LIMIT 25
       ;
```

```
                operation_id                 |           description            |    kind     
---------------------------------------------+----------------------------------+-------------
 deleteCoreV1CollectionNamespacedPodTemplate | delete collection of PodTemplate | PodTemplate
(1 row)
```

# API Reference and feature documentation
* [Kubernetes API Reference Docs](https://kubernetes.io/docs/reference/kubernetes-api/)
* [Kubernetes API: v1.18 Pod Template v1 core Delete Collection](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.18/#delete-collection-podtemplate-v1-core)
* [client-go - PodTemplate: DeleteCollection](https://github.com/kubernetes/client-go/blob/master/kubernetes/typed/core/v1/podtemplate.go)

# The mock test
## Test outline
1. Create a set of 3 Pod Templates with a static label within a Namespace.
2. Confirm that all 3 Pod Templates with the label in the Namespace are created.
3. Delete the set of Namespaced Pod Templates with a label via DeleteCollection.
4. Confirm that all Pod Templates with the label have been deleted.

## Test the functionality in Go
```go
package main

import (
  // "encoding/json"
  "fmt"
  "time"
  "flag"
  "os"
  v1 "k8s.io/api/core/v1"
  // "k8s.io/client-go/dynamic"
  // "k8s.io/apimachinery/pkg/runtime/schema"
  metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
  "k8s.io/client-go/kubernetes"
  // "k8s.io/apimachinery/pkg/types"
  "k8s.io/client-go/tools/clientcmd"
)

func main() {
  // uses the current context in kubeconfig
  kubeconfig := flag.String("kubeconfig", fmt.Sprintf("%v/%v/%v", os.Getenv("HOME"), ".kube", "config"), "(optional) absolute path to the kubeconfig file")
  flag.Parse()
  config, err := clientcmd.BuildConfigFromFlags("", *kubeconfig)
  if err != nil {
      fmt.Println(err)
      return
  }
  // make our work easier to find in the audit_event queries
  config.UserAgent = "live-test-writing"
  // creates the clientset
  ClientSet, _ := kubernetes.NewForConfig(config)
  // DynamicClientSet, _ := dynamic.NewForConfig(config)
  // podResource := schema.GroupVersionResource{Group: "", Version: "v1", Resource: "pods"}

  // TEST BEGINS HERE

  fmt.Println("create a set of Pod templates with a label")
  podTemplateTestNames := [3]string{"test-podtemplate-1", "test-podtemplate-2", "test-podtemplate-3"}

  for _, podTemplateTestName := range podTemplateTestNames {
    _, err = ClientSet.CoreV1().PodTemplates("default").Create(&v1.PodTemplate{
        ObjectMeta: metav1.ObjectMeta{
        Name: podTemplateTestName,
        Labels: map[string]string{
          "type": "Testing"},
        },
        Template: v1.PodTemplateSpec{
        Spec: v1.PodSpec{
          Containers: []v1.Container{
            {Name: "nginx", Image: "nginx"},
          },
        },
      },
    })

    if err != nil {
      fmt.Println("[error]", err)
      return
    }
  fmt.Println("[status] created", podTemplateTestName)
  }

  fmt.Println("get a list of Pod templates with a label in the default namespace")
  podTemplateList, err := ClientSet.CoreV1().PodTemplates("default").List(metav1.ListOptions{
    LabelSelector: "type=Testing"})

  fmt.Println("check that there are 3 Pod templates found")
  fmt.Println("[status] Found", len(podTemplateList.Items), "Pod templates")
  if len(podTemplateList.Items) != 3 {
    fmt.Println("[error] Incorrect number of Pod templates found")
    return
  }

  fmt.Println("delete Collection of Pod templates for the label in the default namespace")
  _  = ClientSet.CoreV1().PodTemplates("default").DeleteCollection(&metav1.DeleteOptions{}, metav1.ListOptions{
         LabelSelector: "type=Testing"})
  fmt.Println("[status] DeleteCollection processed")

  time.Sleep(2 * time.Second)

  fmt.Println("get a list of Pod templates with a label in the default namespace")
  podTemplateList, err = ClientSet.CoreV1().PodTemplates("default").List(metav1.ListOptions{
    LabelSelector: "type=Testing"})

  fmt.Println("check that there are no Pod templates found")
  fmt.Println("[status] Found", len(podTemplateList.Items), "Pod templates")
  if len(podTemplateList.Items) != 0 {
    fmt.Println("[error] Incorrect number of Pod templates found")
    return
  }

  // TEST ENDS HERE

  fmt.Println("[status] complete")

}
```

```
create a set of Pod templates with a label
[status] created test-podtemplate-1
[status] created test-podtemplate-2
[status] created test-podtemplate-3
get a list of Pod templates with a label in the default namespace
check that there are 3 Pod templates found
[status] Found 3 Pod templates
delete Collection of Pod templates for the label in the default namespace
[status] DeleteCollection processed
get a list of Pod templates with a label in the default namespace
check that there are no Pod templates found
[status] Found 0 Pod templates
[status] complete
```

# Verifying increase in coverage with APISnoop
Discover useragents:

```
select distinct useragent from audit_event where bucket='apisnoop' and useragent not like 'kube%' and useragent not like 'coredns%' and useragent not like 'kindnetd%' and useragent like 'live%';
```

```
     useragent     
-------------------
 live-test-writing
(1 row)
```

List endpoints hit by the test:

```
select * from endpoints_hit_by_new_test where useragent like 'live%';
```

```
     useragent     |                operation_id                 | hit_by_ete | hit_by_new_test 
-------------------+---------------------------------------------+------------+-----------------
 live-test-writing | createCoreV1NamespacedPodTemplate           | t          |               3
 live-test-writing | deleteCoreV1CollectionNamespacedPodTemplate | f          |               1
 live-test-writing | listCoreV1NamespacedPodTemplate             | t          |               2
(3 rows)
```

Display endpoint coverage change:

```
select * from projected_change_in_coverage;
```

```
   category    | total_endpoints | old_coverage | new_coverage | change_in_number 
---------------+-----------------+--------------+--------------+------------------
 test_coverage |             445 |          188 |          189 |                1
(1 row)
```

# Final notes
If a test with these calls gets merged, ****test coverage will go up by 1 points****

This test is also created with the goal of conformance promotion.

### PR DESCRIPTION
**Pull request by riaankleinhans**
What type of PR is this?
/kind cleanup

What this PR does / why we need it:
This PR adds a test to test the following untested endpoints:
deleteCoreV1CollectionNamespacedPodTemplate

Which issue(s) this PR fixes:
Fixes #90912

Special notes for your reviewer:
Adds +1 endpoint test coverage (good for conformance)

Does this PR introduce a user-facing change?:
NONE

Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:
NONE

/sig testing

### CODE REVIEW

File: test/e2e/common/podtemplates.go [hunk 1]
```diff
@@ -99,4 +100,79 @@ var _ = ginkgo.Describe("[sig-node] PodTemplates", func() {
...
+		for _, podTemplateName := range podTemplateNames {
+			_, err := f.ClientSet.CoreV1().PodTemplates(f.Namespace.Name).Create(context.TODO(), &v1.PodTemplate{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:   podTemplateName,
+					Labels: map[string]string{"podtemplate-set": "true"},
+				},
+				Template: v1.PodTemplateSpec{
+					Spec: v1.PodSpec{
+						Containers: []v1.Container{
+							{Name: "nginx", Image: "nginx"},
```
Comment (spiffxp) [inline test/e2e/common/podtemplates.go:119]: Let use agnhost instead

Comment (hh) [inline test/e2e/common/podtemplates.go:119]: there is a agnhost image, and it has a pause command.
Let's update that here.

File: test/e2e/common/podtemplates.go [hunk 2]
```diff
@@ -99,4 +100,79 @@ var _ = ginkgo.Describe("[sig-node] PodTemplates", func() {
...
+
+		ginkgo.By("delete collection of pod templates")
+		// delete collection
+
+		framework.Logf("requesting DeleteCollection of pod templates")
+		err = f.ClientSet.CoreV1().PodTemplates(f.Namespace.Name).DeleteCollection(context.TODO(), metav1.DeleteOptions{}, metav1.ListOptions{
+			LabelSelector: "podtemplate-set=true"})
+		framework.ExpectNoError(err, "failed to delete all pod templates")
+
+		ginkgo.By("get a list of pod templates with a label in the current namespace")
```
Comment (hh) [inline test/e2e/common/podtemplates.go:156]: Needs to use the new watch tooling OR a pool to catch the delete.

File: test/e2e/common/podtemplates.go [hunk 3]
```diff
@@ -99,4 +100,79 @@ var _ = ginkgo.Describe("[sig-node] PodTemplates", func() {
...
+		errMsg := "found " + strconv.Itoa(podTemplateListItemsCount) + " pod templates when " + strconv.Itoa(podTemplateNamesCount) + " pod templates where expected"
+		logMsg := "found " + strconv.Itoa(podTemplateListItemsCount) + " pod templates"
+
+		foundCreatedSet := true
+		if podTemplateListItemsCount != podTemplateNamesCount {
+			foundCreatedSet = false
+		} else {
+			framework.Logf(logMsg)
+		}
+		framework.ExpectEqual(foundCreatedSet, true, errMsg)
```
Comment (spiffxp) [inline test/e2e/common/podtemplates.go:146]: I'm not sure we need 11 lines to express when 1 would do? They make it harder to understand what this test is doing at a glance
```suggestion
		framework.ExpectEqual(len(podTemplateList.Items), len(podTemplateNamesCount), "looking for expected number of pod templates")
```

File: test/e2e/common/podtemplates.go [hunk 4]
```diff
@@ -99,4 +100,79 @@ var _ = ginkgo.Describe("[sig-node] PodTemplates", func() {
...
+		errMsg = "found " + strconv.Itoa(podTemplateListItemsCount) + " pod templates when zero pod templates where expected"
+		logMsg = "found " + strconv.Itoa(podTemplateListItemsCount) + " pod templates"
+
+		foundCreatedSet = false
+		if podTemplateListItemsCount != 0 {
+			foundCreatedSet = true
+		} else {
+			framework.Logf(logMsg)
+		}
+		framework.ExpectEqual(foundCreatedSet, false, errMsg)
```
Comment (spiffxp) [inline test/e2e/common/podtemplates.go:174]: same here
```suggestion
		framework.ExpectEqual(len(podTemplateList.Items), 0, "pod templates should all be deleted")
```

**Review by spiffxp — COMMENTED** [review]
there may be better error message to use, but this seems like more boilerplate than it needs to be

File: test/e2e/common/podtemplates.go (hunk 1 above)
Comment (heyste) [reply test/e2e/common/podtemplates.go]: [Fixed with](https://github.com/ii/kubernetes/commit/7772a42114f271296cd3817b58e5e025aa65905d#diff-a73abb634e908184c070916412d4e060R120) `Image: imageutils.GetE2EImage(imageutils.Agnhost)`

Comment (heyste) [reply test/e2e/common/podtemplates.go]: Using `Image: imageutils.GetE2EImage(imageutils.Agnhost)` now

File: test/e2e/common/podtemplates.go (hunk 3 above)
Comment (heyste) [reply test/e2e/common/podtemplates.go]: [Fixed](https://github.com/kubernetes/kubernetes/commit/758e3c76a4fd8b037948713f032f47545f1b0664)

File: test/e2e/common/podtemplates.go (hunk 4 above)
Comment (heyste) [reply test/e2e/common/podtemplates.go]: [Fixed](https://github.com/kubernetes/kubernetes/commit/758e3c76a4fd8b037948713f032f47545f1b0664)

File: test/e2e/common/podtemplates.go (hunk 2 above)
Comment (heyste) [reply test/e2e/common/podtemplates.go:149]: [Fixed:](https://github.com/ii/kubernetes/commit/d373bef4e4aa8d3331f82c6c8bd4404c746e1d06) using `wait.PollImmediate` to check that the `DeleteCollection` as been completed

**Comment by Riaankl** [comment]
@spiffxp 
@johnbelamaric 
Can we get a/lgtm?

File: test/e2e/common/podtemplates.go (hunk 2 above)
Comment (spiffxp) [reply test/e2e/common/podtemplates.go:149]: I'm not sure wrapping the delete call in a poll loop is what we discussed.  Is DeleteCollection a synchronous call? As in, does it block until the entire collection has been deleted?  I would have expected something like polling the list of pod templates post-delete until it goes to zero (or times out)

**Review by spiffxp — APPROVED** [review]
/approve
/lgtm

**Comment by Riaankl** [comment]
/assign @johnbelamaric
Can we get approval please

**Pull request merged**